### PR TITLE
Support TypeScript 3.0

### PR DESCRIPTION
**Pull request by staltz**
- [x] There exists an issue discussing the need for this PR, #815
- [x] I added new tests for the issue I fixed or built
- [x] I used `make commit` instead of `git commit`
- [x] Update /run
- [x] Update /rxjs-run, /most-run
- [x] Update /isolate
- [ ] release candidate /isolate
- [ ] Update onionify
- [ ] Update /dom, /html, /http, /history, /time
- [ ] Release new /run
- [ ] Use new /run in /most-run and /rxjs-run
- [ ] Release new /most-run and /rxjs-run

### CODE REVIEW

File: run/src/types.ts [hunk 1]
```diff
@@ -18,36 +18,42 @@ export interface DevToolEnabledSource {
...
-  sources: So;
-  run(sinks: Si): DisposeFunction;
+export interface Engine<D extends Drivers> {
+  sources: Sources<D>;
+  run<M extends Main>(sinks: Sinks<M>): DisposeFunction;
   dispose(): void;
 }
+
+export type MatchingMain<D extends Drivers, M extends Main> = Main & {
+  (so?: Partial<Sources<D>>): Sinks<M>;
```
Comment (jvanbruegge) [inline run/src/types.ts]: What exactly does this do?

File: run/test/setup.ts [hunk 2]
```diff
@@ -47,52 +48,44 @@ describe('setup', function() {
...
-      obj: Stream<object>;
-    };
-
-    function app(sources: Sources) {
+    function app(sources: {str: Stream<string>; obj: Stream<object>}) {
       return {
         str: sources.str.take(1).startWith('a'), // good
         // str: sources.obj.mapTo('good'), // good
-        // strTYPO: sources.str.take(1).startWith('a'), // bad
+        // strTYPO: xs.of('a'), // good (but in older versions was bad)
```
Comment (jvanbruegge) [inline run/test/setup.ts]: Shouldnt this be bad if the key is not in the drivers object?

File: run/test/setup.ts [hunk 3]
```diff
@@ -152,12 +141,12 @@ describe('setup', function() {
       other: Stream<number>;
     };
 
-    function app(sources: TestSources) {
+    function app(so: TestSources) {
```
Comment (jvanbruegge) [inline run/test/setup.ts]: Why this change?

File: run/test/setup.ts [hunk 4]
```diff
@@ -47,52 +48,44 @@ describe('setup', function() {
...
       num: numberWriteOnlyDriver,
       obj: objectReadOnlyDriver,
-    });
+    };
+    setup(app, drivers);
   });
 
   it('should type-check keyof sources and sinks, supporting interfaces', function() {
-    interface Sources {
+    interface MaSources {
```
Comment (jvanbruegge) [inline run/test/setup.ts]: Why this change?

File: run/src/types.ts [hunk 5]
```diff
@@ -18,36 +18,42 @@ export interface DevToolEnabledSource {
...
-
 export type DisposeFunction = () => void;
 
-export interface CycleProgram<So extends Sources, Si extends Sinks> {
-  sources: So;
-  sinks: Si;
+export type Drivers = {
+  [name: string]: (...args: Array<any>) => any;
+};
+export type Main = (...args: Array<any>) => any;
```
Comment (jvanbruegge) [inline run/src/types.ts]: Shouldnt main just take the sources?

File: run/src/types.ts [hunk 6]
```diff
@@ -18,36 +18,42 @@ export interface DevToolEnabledSource {
...
-  [P in keyof (So & Si)]: Driver<Si[P], So[P]>
-};
-
 export type DisposeFunction = () => void;
 
-export interface CycleProgram<So extends Sources, Si extends Sinks> {
-  sources: So;
-  sinks: Si;
+export type Drivers = {
+  [name: string]: (...args: Array<any>) => any;
```
Comment (jvanbruegge) [inline run/src/types.ts]: Shouldnt a driver just take a stream?

File: run/src/types.ts (hunk 1 above)
Comment (staltz) [reply run/src/types.ts]: MatchingMain is basically the "set of main functions that can possibly match the given drivers object" and MatchingDrivers is the "set of driver objects that can possibly match the given main function".

File: run/test/setup.ts (hunk 2 above)
Comment (staltz) [reply run/test/setup.ts]: So, there are a couple of real use cases where I take a ready `main` function and plug it with my drivers, and I choose to ignore a certain sink because I'm not going to use it.

That said, I think you have a point, maybe we could support that case by forcing the user to give a noop driver for that channel. But the other way around (drivers provide some sources but `main` never use them) doesn't need this kind of check.

File: run/test/setup.ts (hunk 3 above)
Comment (staltz) [reply run/test/setup.ts]: This was probably testing frustration, I can revert.

File: run/test/setup.ts (hunk 4 above)
Comment (staltz) [reply run/test/setup.ts]: Testing frustration, I can revert

File: run/src/types.ts (hunk 5 above)
Comment (staltz) [reply run/src/types.ts]: This `Main` type for now is quite broad, but it makes sense when used together with all the other stuff. Also I'm explicitly not defining Main as `So => Si` neither a Driver as `Si => So`. Instead, I'm defining Sources as ReturnType<Drivers> and Sinks as ReturnType<Main>

File: run/src/types.ts (hunk 6 above)
Comment (staltz) [reply run/src/types.ts]: See above

File: run/test/setup.ts (hunk 2 above)
Comment (jvanbruegge) [reply run/test/setup.ts]: Yeah, i think everything returned by main should be in the drivers object, but not the other way round

**Comment by staltz** [comment]
Updated the commit to be much more conservative. It now has no breaking changes, and minimal changes.

File: most-run/src/index.ts [hunk 7]
```diff
@@ -1,21 +1,50 @@
 import {Stream} from 'xstream';
 import * as most from 'most';
 import {Stream as MostStream} from 'most';
-import {setAdapt} from '@cycle/run/lib/adapt';
+import {setAdapt} from '../../run/src/adapt';
```
Comment (staltz) [inline most-run/src/index.ts]: **TODO** replace with a new released version of /run

File: most-run/src/index.ts [hunk 8]
```diff
@@ -1,21 +1,50 @@
...
   DisposeFunction,
-  Driver,
-  FantasyObservable,
-  Sources,
-  Sinks,
-  CycleProgram,
-} from '@cycle/run';
+  Drivers,
+  Main,
+} from '../../run/src/index';
```
Comment (staltz) [inline most-run/src/index.ts]: **TODO** replace with a new released version of /run

File: most-run/src/index.ts (hunk 7 above)
Comment (jvanbruegge) [inline most-run/src/index.ts]: I know you need that for testing, but you could use `pnpm link` instead.
This wont work for release and could be forgotten

File: most-run/src/index.ts (hunk 8 above)
Comment (jvanbruegge) [inline most-run/src/index.ts]: Same

File: most-run/test/run.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    const dispose = run(app, {other: driver});
+    assert.strictEqual(typeof dispose, 'function');
+    setTimeout(() => {
+      sinon.assert.calledOnce(spy);
+    });
+    dispose();
+  });
+
+  // Skipped until we make this test not brittle with async timing
```
Comment (jvanbruegge) [inline most-run/test/run.ts]: Outdated comment

File: most-run/test/setup.ts [hunk 10]
```diff
@@ -39,15 +38,15 @@ describe('setup', function() {
       other: Stream<string>;
     };
 
-    function app(sources: MySources): MySinks {
+    function app(_sources: MySources): MySinks {
```
Comment (jvanbruegge) [inline most-run/test/setup.ts]: Why the renaming?

File: rxjs-run/src/index.ts [hunk 11]
```diff
@@ -2,21 +2,50 @@
 import 'symbol-observable';
 import {Stream} from 'xstream';
 import {from, Observable} from 'rxjs';
-import {setAdapt} from '@cycle/run/lib/adapt';
+import {setAdapt} from '../../run/src/adapt';
```
Comment (jvanbruegge) [inline rxjs-run/src/index.ts]: Same as most-run

File: most-run/test/setup.ts (hunk 10 above)
Comment (staltz) [reply most-run/test/setup.ts]: Because of variable name shadowing, see some lines below this

File: rxjs-run/test/run.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+  });
+
+  it('should throw if second argument is an empty object', function() {
+    assert.throws(() => {
+      (run as any)(() => {}, {});
+    }, /Second argument given to Cycle must be an object with at least one/i);
+  });
+
+  it('should return a dispose function', function() {
+    const sandbox = sinon.sandbox.create();
```
Comment (jvanbruegge) [inline rxjs-run/test/run.ts]: Is deprecated, use `sinon.createSandbox()` instead

File: rxjs-run/test/setup.ts [hunk 13]
```diff
@@ -38,9 +38,9 @@ describe('setup', function() {
       other: Observable<string>;
     };
 
-    function app(sources: MySources): MySinks {
+    function app(_sources: MySources): MySinks {
```
Comment (jvanbruegge) [inline rxjs-run/test/setup.ts]: Why rename here?

File: most-run/test/setup.ts (hunk 10 above)
Comment (jvanbruegge) [reply most-run/test/setup.ts]: I see

File: rxjs-run/test/run.ts (hunk 12 above)
Comment (staltz) [reply rxjs-run/test/run.ts]: Hmm, the types don't allow me to write sinon.createSandbox(). Maybe you updated sinon's version in the pnpm PR? In that case we can do a merge of those two branches, whichever gets merged second adapts accordingly.

File: rxjs-run/test/setup.ts (hunk 13 above)
Comment (staltz) [reply rxjs-run/test/setup.ts]: variable name shadowing

File: rxjs-run/test/run.ts (hunk 12 above)
Comment (jvanbruegge) [reply rxjs-run/test/run.ts]: Yeah, I updated all dependencies

File: most-run/src/index.ts (hunk 7 above)
Comment (staltz) [reply most-run/src/index.ts]: How could we make it work for CI for now?

**Comment by staltz** [comment]
I agree

**Comment by jvanbruegge** [comment]
I've rebased onto the current master

**Comment by jvanbruegge** [comment]
closing in favor of #844
